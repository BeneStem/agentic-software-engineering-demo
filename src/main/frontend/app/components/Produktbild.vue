<template>
  <div class="erk_finden__produktbild">
    <picture v-if="defaultImage">
      <source
        v-for="pictureSource in pictureSources"
        :key="pictureSource.media"
        :media="pictureSource.media"
        :type="pictureSource.type"
        :srcset="pictureSource.srcset"
      />
      <img :src="defaultImage" :srcset="defaultImageSrcSet" :alt="altText" loading="lazy" @error="onError" />
    </picture>
  </div>
</template>

<script setup lang="ts">
import { ref } from 'vue';
import { PictureSource } from '@blume2000/design-system/dist/models/PictureSource';

interface Props {
  url: string;
  altText: string;
}

const props = defineProps<Props>();

const defaultDpr = 2;
const allDprs = [1, 2, 3];
const defaultImage = ref('');
const defaultImageSrcSet = ref('');

const finalUrl = (width: number, height: number, dpr: number) =>
  props.url.replace('%w%', width.toString()).replace('%h%', height.toString()).replace('%d%', dpr.toString());

const finalSrcSet = (width: number, height: number): string => {
  if (!width || !height) {
    return '';
  }
  return allDprs.map((dpr: number) => `${finalUrl(width, height, dpr)} ${dpr}x`).join(',');
};

defaultImage.value = finalUrl(328, 328, defaultDpr);
defaultImageSrcSet.value = finalSrcSet(328, 328);

const pictureSources: PictureSource[] = [
  {
    srcset: finalSrcSet(401, 401),
    media: '(min-width: 361px)',
    type: 'image/jpeg',
  },
];

const onError = () => {
  defaultImage.value = 'https://assets.ecom.blume2000.de/images/produkt_platzhalter.png';
  defaultImageSrcSet.value = '';
};
</script>
