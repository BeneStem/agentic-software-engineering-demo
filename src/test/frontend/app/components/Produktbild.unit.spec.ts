import { mount } from '@vue/test-utils';
import Produktbild from '@/main/frontend/app/components/Produktbild.vue';

describe('Produktbild.vue', () => {
  it('Rendert Produktbild', () => {
    // Given
    const wrapper = mount(Produktbild, {
      props: {
        url: 'https://res.cloudinary.com/blume2000/image/upload/c_fill,dpr_%d%,f_auto,h_%h%,q_auto,w_%w%/v1/prod/produkte/PRIMAER/20003385',
        altText: 'Alt text',
      },
    });

    // Then
    expect(wrapper.element).toMatchSnapshot();
  });

  it('Zeigt Platzhalter, wenn das Produktbild nicht geladen werden kann', async () => {
    // Given
    const wrapper = mount(Produktbild, {
      props: {
        url: 'gibts nicht',
        altText: 'Alt text',
      },
    });

    await wrapper.find('img').trigger('error');

    // Then
    expect(wrapper.find('img').attributes()['src']).toEqual(
      'https://assets.ecom.blume2000.de/images/produkt_platzhalter.png'
    );
  });
});
