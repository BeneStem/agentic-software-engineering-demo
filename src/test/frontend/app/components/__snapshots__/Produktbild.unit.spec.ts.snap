// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Produktbild.vue Rendert Produktbild 1`] = `
<div
  class="erk_finden__produktbild"
>
  <picture>
    
    <source
      media="(min-width: 361px)"
      srcset="https://res.cloudinary.com/blume2000/image/upload/c_fill,dpr_1,f_auto,h_401,q_auto,w_401/v1/prod/produkte/PRIMAER/20003385 1x,https://res.cloudinary.com/blume2000/image/upload/c_fill,dpr_2,f_auto,h_401,q_auto,w_401/v1/prod/produkte/PRIMAER/20003385 2x,https://res.cloudinary.com/blume2000/image/upload/c_fill,dpr_3,f_auto,h_401,q_auto,w_401/v1/prod/produkte/PRIMAER/20003385 3x"
      type="image/jpeg"
    />
    
    <img
      alt="Alt text"
      loading="lazy"
      src="https://res.cloudinary.com/blume2000/image/upload/c_fill,dpr_2,f_auto,h_328,q_auto,w_328/v1/prod/produkte/PRIMAER/20003385"
      srcset="https://res.cloudinary.com/blume2000/image/upload/c_fill,dpr_1,f_auto,h_328,q_auto,w_328/v1/prod/produkte/PRIMAER/20003385 1x,https://res.cloudinary.com/blume2000/image/upload/c_fill,dpr_2,f_auto,h_328,q_auto,w_328/v1/prod/produkte/PRIMAER/20003385 2x,https://res.cloudinary.com/blume2000/image/upload/c_fill,dpr_3,f_auto,h_328,q_auto,w_328/v1/prod/produkte/PRIMAER/20003385 3x"
    />
  </picture>
</div>
`;
